legacy/firmware/crypto.c

```diff
@@ -471,19 +471,16 @@ int cryptoMultisigFingerprint(const MultisigRedeemScriptType *multisig,
 
   SHA256_CTX ctx = {0};
   sha256_Init(&ctx);
-  sha256_Update(&ctx, (const uint8_t *)&(multisig->m), sizeof(uint32_t));
-  sha256_Update(&ctx, (const uint8_t *)&(pubkeys_order), sizeof(uint32_t));
+  SHA256_UPDATE_INT(&ctx, multisig->m, uint32_t);
+  SHA256_UPDATE_INT(&ctx, pubkeys_order, uint32_t);
   for (uint32_t i = 0; i < n; i++) {
-    sha256_Update(&ctx, (const uint8_t *)&(pubnodes[i]->depth),
-                  sizeof(uint32_t));
-    sha256_Update(&ctx, (const uint8_t *)&(pubnodes[i]->fingerprint),
-                  sizeof(uint32_t));
-    sha256_Update(&ctx, (const uint8_t *)&(pubnodes[i]->child_num),
-                  sizeof(uint32_t));
+    SHA256_UPDATE_INT(&ctx, pubnodes[i]->depth, uint32_t);
+    SHA256_UPDATE_INT(&ctx, pubnodes[i]->fingerprint, uint32_t);
+    SHA256_UPDATE_INT(&ctx, pubnodes[i]->child_num, uint32_t);
     sha256_Update(&ctx, pubnodes[i]->chain_code.bytes, 32);
     sha256_Update(&ctx, pubnodes[i]->public_key.bytes, 33);
```

Comment:
What about these two lines? Since we agree that code should use actual sizes of the fields / data types, instead of assuming a particular size, we should replace `32` with `pubnodes[i]->chain_code.size` or `sizeof(pubnodes[i]->chain_code.bytes)` and similarly `33`.

Reply:
BTW, there is probably a lot of similar cases, for example in firmware/reset.c `sha256_Update(&ctx, int_entropy, 32)` should be `sha256_Update(&ctx, int_entropy, sizeof(int_entropy))`, but maybe that's out of scope of this PR.

Reply:
We can introduce similar macros named HASHER_UPDATE_BYTES where we provide expected size in bytes and macro will do the Static_assert.

Reply:
Added `HASHER_UPDATE_BYTES` / `SHA256_UPDATE_BYTES` in aa16e91, please take a look.